wxGUI: Fix key binding issues on mac

nilason opened the pull request:
Fixes #785

This PR will replace #792.

The main problem with stock id buttons is that they all come with a key binding. E.g. Close, Clear and Copy have the same key binding ctrl+c and causes problems at least for mac. This is the most generalised approach I could come up with to remedy this issue (using subclassed versions of Button: ClearButton, CancelButton, ApplyButton).

I also made some effort to implement ESC to close dialog,  on the dialogs I have modified.

Although this works fine for mac on wxPython 4.0.7.post2 as well as 4.1.0, it needs thorough testing on Linux and Windows too.

Note: for module dialogs (forms.py) and Map calculator I have replaced ctrl+c to copy selected text, not to copy the whole command. The latter is, I believe, a very unconventional use of ctrl+c on any platform, but if there are objections to this it may be reimplemented with conditionals.

petrasovaa commented on gui/wxpython/gui_core/wrap.py:
```diff
@@ -178,6 +179,42 @@ def SetToolTip(self, tip):
+    def __init__(self, *args, **kwargs):
+        Button.__init__(self, *args, **kwargs)
+        self.SetId(wx.ID_CANCEL)
+        if sys.platform == "darwin" and not CheckWxVersion([4, 1, 0]):
```
Why special checks for wx 4.1.0 in CancelButton and not in other buttons?

nilason replied:
This particular issue with `Cancel` was addressed with https://github.com/wxWidgets/wxWidgets/pull/1131 which went to wxWidgets 3.1.4 (and thus wxPython 4.1). This check corrects this for wxPython <4.1 (and restores default label `&Cancel` for later versions).

petrasovaa commented on gui/wxpython/modules/import_export.py:
```diff
@@ -116,14 +116,12 @@ def __init__(self, parent, giface, itype,
         #
         # cancel
-        self.btn_close = Button(parent=self.panel, id=wx.ID_CLOSE)
+        self.btn_close = CancelButton(parent=self.panel)
         self.btn_close.SetToolTip(_("Close dialog"))
```
This changes is from "Close" to "Cancel", I would think "Close" would be often more appropriate. Maybe we need a CloseButton? This is on other places as well.

nilason replied:
Maybe a bad move for me to mix addressing the original issue (with key bindings) with GUI changes from `Close` to `Cancel`!

This PR was intended to be a step in (and prerequisite for) a more general reorganisation of buttons as I mentioned earlier https://github.com/OSGeo/grass/pull/744#issuecomment-663977673. The rationale for the change of labels is to make a consistent GUI: both within GRASS and compared to common practice in various applications and OSs. Now with GRASS the two labels are used variably. In my experience, in most environments you **close** a document/file, but **cancel** a dialog window.

wenzeslaus replied:
> In my experience, in most environments you **close** a document/file, but **cancel** a dialog window.

I think you *close* a document/file/book and *cancel* an operation. If there is an operation behind a dialog, let's say import, Cancel button should cancel the import operation. If the dialog is used to run an operation (import file), but is open even when the operation was started (or finished; file was imported), Close button is better since you are closing the dialog/window, nothing is being canceled (the operation was done already; file was imported). Perhaps the issue is window versus dialog terminology or perhaps you are already using this rule (I didn't check).

wenzeslaus replied:
I find the "by opening dialog window you already started an operation" part convincing. I'll think about it more for the Cancel/Close PR.

petrasovaa reviewed (commented):
I tested it on Ubuntu and it seems to work (I haven't notice any changes in behavior). Please look at my comments, but otherwise we could merge it. I would wait with backport though.

nilason commented:
Added a CloseButton wrap class and reverted the `Cancel` buttons to `Close` buttons.
That change to Cancel buttons should be a separate issue/PR and I wasn't consistent either.

petrasovaa commented:
> Added a CloseButton wrap class and reverted the `Cancel` buttons to `Close` buttons.
> That change to Cancel buttons should be a separate issue/PR and I wasn't consistent either.

Thanks. You are right about the cancel/close issue, but changing it in a separate PR will be better. I plan on getting back to this soonish.

petrasovaa reviewed (approved):
Looks good! I wouldn't backport this just yet, would be better to test this little more.

nilason commented:
Great, thank you!
I agree regarding postponing backport.

infrastation commented:
One of these changes breaks GUI on Linux, please make sure issue #1156 has been resolved before backporting.

neteler commented:
Can anyone with a Mac pls check the state here: backport-ready or not yet?

nilason commented:
> Can anyone with a Mac pls check the state here: backport-ready or not yet?

I believe it should be safe, at least on Mac it is stable. The only issue reported so far is fixed with #1160 (which in any case should be backported **after** this one).
But I need to pass the question on ready-for-backport to a non-mac user.

petrasovaa commented:
I would say it's ready for backport.
